redundant-bmc/src/redundancy.cpp

```
@@ -177,4 +179,109 @@ std::string getFailoversNotAllowedDescription(FailoversNotAllowedReason reason)
 }
 
 } // namespace fona
+
+namespace fo_blocked
+{
+
+Reason getFailoverBlockedReason(const Input& input)
+{
+    if (input.siblingHeartbeat)
+    {
+        if (!input.redundancyEnabled)
+        {
+            return Reason::redundancyNotEnabled;
+        }
+        else if (input.failoversNotAllowed)
+        {
+            // Don't block a failover even if the failover is not allowed when:
+            //  1. the force option was given on the start failover cmd, or
+            //  2. the active BMC is Quiesced.
+            if (input.forceOption)
+            {
+                // Trace it but don't block it.
+                lg2::info(
+                    "The failover 'Force' option is set while failovers are not allowed");
+            }
+            else if (input.siblingState == BMCState::Quiesced)
+            {
+                // If the active BMC is quiesced, it may be stuck in
+                // failovers-not-allowed so don't block it, just trace it.
+                lg2::info(
+                    "The sibling BMC is quiesced while failovers are not allowed");
+            }
+            else
+            {
+                return Reason::failoversNotAllowed;
+            }
+        }
+        else if (input.syncInProgress)
```

Do we have situations where a bad primary BMC could cause this to be permanently set? Is there a timeout? This one worries me the most because it could occur for a variety of reasons and it would be fairly unexpected for the HMC or PHYP to get this response. Do we tell them to retry? In 30s? 1 minute?

---

@RameshIyyar Do you have any worries here about a full sync never completing?  How my doFullSync() code is written now it waits on a propertiesChanged signal so it can't time out.

---

It will eventually complete—either successfully or with a failure—but the time it takes is uncertain due to our retry mechanism.

---

While thinking further, I was wondering—can we trigger some kind of event to HMC/PHYP to retry the failover, instead of relying on a timeout? If the timeout mechanism is straightforward/simple, that’s fine. I just want to confirm whether there’s any scenario where they might stop retrying altogether.


> Do we have situations where a bad primary BMC could cause this to be permanently set?

Also, I’d like to clarify @geissonator’s question — here we are checking the Full Sync status of the Passive BMC, not the Active one.

---

Most of the checks here just prevent a failover from starting when it shouldn't have been called in the first place:
  failover in progress/imminent <---- FIP will be on D-Bus, maybe redfish? as you suggested maybe we could return a special error so they'd know they don't need to try again
    redundancyNotEnabled,  <--- shouldn't have been called
    fullSyncInProgress,           <--- probably full sync also in progress on active, so FOAllowed = false
    failoversNotAllowed,         <--- shouldn't have been called
    siblingDeadButRedundancyNotEnabled,  <------- shouldn't have been called because RedEnabled = false
    notAtReady,            <-- they could check for this and not call
    bmcNotPassive        < -- shouldn't have been called

I assume they'll get events when RedEnabled or FOAllowed changes.